src/arena.c

```diff
@@ -156,12 +156,13 @@ arena_stats_merge(tsdn_t *tsdn, arena_t *arena, unsigned *nthreads,
 	astats->tcache_stashed_bytes = 0;
 	malloc_mutex_lock(tsdn, &arena->tcache_ql_mtx);
 	cache_bin_array_descriptor_t *descriptor;
+	tcache_t *tcache = tsd_tcachep_get(tsdn_tsd(tsdn));
```

Comment:
This needs some tweaking: the list below is chaining the tcache->bins from all threads assigned to this arena, therefore each may have their own tcache_max setting. The reason we have the cache bins chained, instead of tcache, is to avoid a circular dependency, i.e. arena should not be tcache aware since it's lower level, and tcache is arena aware already. The cache bins is rather plain old data, so having both tcache and arena aware of it is not a problem.

Two ways we can resolve this:
1) adding the tcache_max into the descriptor type here, so that we know how many bins for each descriptor
https://github.com/jemalloc/jemalloc/blob/b71da25b8a12c2c3f0c10b0811d15a61980186e8/include/jemalloc/internal/cache_bin.h#L137
2) make the for loop below always go all the way up to the max allowed bins, and make sure the disabled bins are handled properly.

Which way is better might also be related to the next PR of the tunable ncached_max. WDYT @guangli-dai ?

Reply:
I think the second approach is better. Yet we do not initialize large bins, which makes the `disabled` check rely on `ncached_max` only. This won't be a problem in this PR, but will be in the upcoming PR that pre-generates `ncached_max` for each bin. Thus, I want to first check whether you think it's fine to init large bins, i.e., bin_ind > SC_NBINS, to fake stack as well?

Reply:
IIUC, you mean doing something similar to:
https://github.com/jemalloc/jemalloc/blob/b71da25b8a12c2c3f0c10b0811d15a61980186e8/src/tcache.c#L690-L699

But for nhbins < bins < max # of tcache bins, is that correct? If so I think that's alright. Please add some assertions to catch those cases though, since if we set it up, it won't crash on access anymore.

Reply:
Sure. Modified the current PR, simply use ncached_max == 0 as the disabled condition for now. Will update 1. the initialization of large bins; 2. the disabled check of large bins; 3. assertions regarding large bins in the next PR (#2530) which enabled ncached_max to be non-zero even with the bin disabled.

Reply:
Thinking about it again, I'm not sure I follow why bins > nhbins need to be set up with fake stacks. The reason for the bins <= nhbins to have it, is because we allow fast path to access them, using the exact same cache_bin_alloc flow w/o checking ncached_max. For the higher bins, can't we simply set ncached_max to the value we want, and set the stack pointers to null? Then in the bin_is_disabled check, the logic should still hold even with null pointers (e.g. empty_position == full_position). That way we still make sure bins > nhbins will crash on access.

Reply:
I think you are right that we can still use empty_position == full_position with them as NULL. This can help keep the next PR neater.